### DISCUSSION TITLE
How to/can we access the state of the ASReview project in real-time without saving to .asreview file

### DISCUSSION THREAD

**Discussion by rohitgarud** [asreview/asreview#1329] [Q&A]
Hello, 
The python API provides a way to access the state from the saved .asreview file. I was wondering whether there is a way to access the state of the review process in real-time so we can perform some analysis. 
Thank you.

**Comment by J535D165** [comment]
Hi, 

Good question. This is possible but experimental and only partially documented.

There are two ways to do this (that pop into my mind):

1. Use the API https://asreview.readthedocs.io/en/latest/simulation_api_example.html. Instead of calling `.review()`, iterate the active learning cycle yourself. See this chunk of code: https://github.com/asreview/asreview/blob/46ccbb4416f525a5c20eb66307f726bfadf0bb76/asreview/review/base.py#L177-L196. 
2. Another option would be to create a custom model (e.g. a classifier) https://github.com/asreview/template-extension-new-model. You let the model access the state file. Now you have access to the state file in every iteration.

The first requires more coding and can't be integrated with the UI or CLI. The second can be used in the UI and CLI, as "extension" models are available there. 

Does this help? If this is a feature more users need, we can make an easy technical interface.

**Reply by rohitgarud** [reply]
Thank you for your quick response. I will try both of these options and update here about my experience.
Regarding the requirement, I think there are various use cases where the state information can be used. For example, performing calculations for the knee method on the recall curve in real-time and checking if the predefined threshold is reached. Currently, we have to download the CSV and run [a script](https://github.com/MaxvanHaastrecht/ASReview-Knee-Method/blob/main/Cormack%20Grossman%20Knee%20Algorithm.ipynb) on it to get the results, which is time-consuming.

**Reply by rohitgarud** [reply]
I have removed the Answer tag for now to have more discussion by the community on this topic.
